src/lib.rs

```diff
@@ -341,6 +341,57 @@ impl<const LEN: usize> AStr<LEN> {
         };
         AStr::<RET_LEN>::from_utf8_array_unchecked(ret_buf)
     }
+
+    pub fn try_from_fmt(display: impl core::fmt::Display) -> Result<Self, core::fmt::Error> {
+        use core::fmt::Write;
+        let mut builder = FmtBuilder::new();
+        write!(builder, "{}", display)?;
+        builder.finalize()
+    }
+}
+
+/// Private type to build an [`AStr`] from anything that can print to an [core::fmt::Write]
+struct FmtBuilder<const LEN: usize> {
+    len: usize,
+    partial: AStr<LEN>,
+}
+
+impl<const LEN: usize> FmtBuilder<LEN> {
+    pub fn new() -> Self {
+        Self {
+            len: 0,
+            partial: AStr::repeat('\0'),
+        }
+    }
+
+    pub fn finalize(self) -> Result<AStr<LEN>, core::fmt::Error> {
+        if self.len == LEN {
+            Ok(self.partial)
+        } else {
+            Err(core::fmt::Error)
+        }
+    }
+}
+
+impl<const LEN: usize> core::fmt::Write for FmtBuilder<LEN> {
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        let s_len = s.len();
+        let offset = self.len;
+
+        self.len = self.len.checked_add(s_len).ok_or(core::fmt::Error)?;
```

Comment:
`self.len` should only be advanced after the bytes were copied

Reply:
IIUC, by contract, the `std::fmt::{Display,Debug,LowerHex,...}` trait implementations should stop trying to print whenever they encounter an error coming from the `Formatter` (which is backed by `FmtBuilder` in our case).

https://doc.rust-lang.org/stable/std/fmt/index.html
> Additionally, the return value of this function is [fmt::Result](https://doc.rust-lang.org/stable/std/fmt/type.Result.html) which is a type alias of [Result](https://doc.rust-lang.org/stable/std/result/enum.Result.html)<(), [std::fmt::Error](https://doc.rust-lang.org/stable/std/fmt/struct.Error.html)>. Formatting implementations should ensure that they propagate errors from the [Formatter](https://doc.rust-lang.org/stable/std/fmt/struct.Formatter.html) (e.g., when calling [write!](https://doc.rust-lang.org/stable/std/macro.write.html)). However, they should never return errors spuriously. That is, a formatting implementation must and may only return an error if the passed-in [Formatter](https://doc.rust-lang.org/stable/std/fmt/struct.Formatter.html) returns an error. This is because, contrary to what the function signature might suggest, string formatting is an infallible operation. This function only returns a result because writing to the underlying stream might fail and it must provide a way to propagate the fact that an error has occurred back up the stack.

I deduce from this that what happens after a write error is more or less undefined, as long as it remains safe and sound. That's why I didn't bother too much with proper error handling, but I'll change that too if you prefer.